evaluation/evaluate.py

```diff
@@ -0,0 +1,186 @@
+        model_config=aoai_config,
+        data_mapping={
+            # Matches qa.jsonl
+            "questions": "question",  # column of data providing input to model
+            "y_test": "truth",  # column of data providing ground truth answer, optional for default metrics
+            # Matches return value of qna()
+            "contexts": "context",  # column of data providing context for each input
+            "y_pred": "answer",  # column of data providing output from model
+        },
+        tracking=False,
+        output_path=destination_dir,
+    )
+
+    eval_results_filename = list(results.artifacts.keys())[0]
+    with open(Path(destination_dir) / eval_results_filename) as f:
+        questions_with_ratings = [json.loads(question_json) for question_json in f.readlines()]
+
+    metrics = {
+        metric_name: {
+            "mean_rating": round(results.metrics_summary[f"mean_{metric_name}"], 2),
+            "pass_count": 0,
+            "pass_rate": 0,
+        }
+        for metric_name in gpt_metrics
+    }
+    total_length = 0
+    max_length = 0
+    min_length = 9999999999
+    total_with_citation = 0
+
+    def passes_threshold(rating):
+        return int(rating) >= 4
+
+    for ind, question_with_rating in enumerate(questions_with_ratings):
+        total_length += len(question_with_rating["answer"])
+        max_length = max(max_length, len(question_with_rating["answer"]))
+        min_length = min(min_length, len(question_with_rating["answer"]))
+        total_with_citation += 1 if re.search(r"\[[^\]]+\]", question_with_rating["answer"]) else 0
+        for metric_name in gpt_metrics:
+            if passes_threshold(question_with_rating[metric_name]):
+                metrics[metric_name]["pass_count"] += 1
+            metrics[metric_name]["pass_rate"] = round(metrics[metric_name]["pass_count"] / (ind + 1), 2)
+    metrics["answer_length"] = {
+        "total": total_length,
+        "mean": round(total_length / len(questions_with_ratings), 2),
+        "max": max_length,
+        "min": min_length,
+    }
+    metrics["answer_has_citation"] = {
+        "total": total_with_citation,
+        "rate": round(total_with_citation / len(questions_with_ratings), 2),
+    }
+
+    # summary statistics
+    with open(Path(destination_dir) / "summary.json", "w") as summary_file:
+        summary_file.write(json.dumps(metrics, indent=4))
+
+    with open(Path(destination_dir) / "parameters.json", "w") as parameters_file:
+        parameters = {
+            "overrides": overrides,
+            "evaluation_gpt_model": gpt_model,
+            "app_gpt_model": "Unknown" if target_url else os.environ["AZURE_OPENAI_CHATGPT_MODEL"],
+            "target_url": target_url if target_url else "localhost/chat",
+        }
+        parameters_file.write(json.dumps(parameters, indent=4))
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.WARNING)
+
+    timestamp = int(time.time())
+
```

Comment:
note from chat with chris:

read in a config.json
if prompttemplate_file, override overrides["prompt_template"] with contents of that file
(relative path? :O )

what if there are JSON characters in the file? it should be fine, but can check.